iroh-net/src/magicsock/node_map/udp_paths.rs

```diff
@@ -0,0 +1,148 @@
+            best_addr,
+            chosen_candidate: None,
+        }
+    }
+
+    /// Returns the current UDP address to send on.
+    ///
+    /// TODO: The goal here is for this to simply return the already known send address, so
+    /// it should be `&self` and not `&mut self`.  This is only possible once the state from
+    /// [`NodeUdpPaths`] is no longer modified from outside.
+    pub(super) fn send_addr(&mut self, now: Instant, have_ipv6: bool) -> UdpSendAddr {
+        self.assign_best_addr_from_candidates_if_empty();
+        match self.best_addr.state(now) {
+            best_addr::State::Valid(addr) => UdpSendAddr::Valid(addr.addr),
+            best_addr::State::Outdated(addr) => UdpSendAddr::Outdated(addr.addr),
+            best_addr::State::Empty => {
+                // No direct connection has been used before.  If we know of any possible
+                // candidate addresses, randomlly try to use one.  This path is most
+                // effective when folks use a NodeAddr with exactly one direct address which
+                // they know to work, effectively like using a traditional socket or quic
+                // endpoint.
+                let addr = self
+                    .chosen_candidate
+                    .and_then(|ipp| self.paths.get(&ipp))
+                    .and_then(|path| path.udp_addr())
+                    .filter(|addr| addr.is_ipv4() || have_ipv6)
+                    .or_else(|| {
+                        // Look for a new candidate in all the known paths.  This may look
+                        // like a RNG use on the hot-path but this is normally invoked at
+                        // most most once at startup.
+                        let addr = self
+                            .paths
+                            .values()
+                            .filter_map(|path| path.udp_addr())
+                            .filter(|addr| addr.is_ipv4() || have_ipv6)
+                            .choose(&mut rand::thread_rng());
+                        self.chosen_candidate = addr.as_ref().map(|addr| IpPort::from(*addr));
+                        addr
+                    });
+                match addr {
+                    Some(addr) => UdpSendAddr::Unconfirmed(addr),
+                    None => UdpSendAddr::None,
+                }
+            }
+        }
+    }
+
+    /// Fixup best_adrr from candidates.
+    ///
+    /// If somehow we end up in a state where we failed to set a best_addr, while we do have
+    /// valid candidates, this will chose a candidate and set best_addr again.  Most likely
+    /// this is a bug elsewhere though.
+    fn assign_best_addr_from_candidates_if_empty(&mut self) {
+        if !self.best_addr.is_empty() {
+            return;
+        }
+
+        // The highest acceptable latency for an endpoint path.  If the latency is higher
+        // then this the path will be ignored.
+        const MAX_LATENCY: Duration = Duration::from_secs(60 * 60);
```

Comment:
this seems rather pointless.. latency of 1 hour?

Reply:
Yes, I don't really know how this is useful either.  Though this code is copy-pasted and I'd really like to avoid doing unrelated changes here.  I do want to carry on fixing up things here though, but in small PRs that each target one thing because my attempt to do a bunch of this together became messy and un-reviewable.